--- ../../applications/bci_visualization/bci_visualization.py	2026-01-20 15:26:14.114811021 +0000
+++ ./bci_visualization.mycopy.py	2026-01-20 15:38:21.031365716 +0000
@@ -29,6 +29,13 @@
 from holohub.color_buffer_passthrough import ColorBufferPassthroughOp
 from holohub.volume_renderer import VolumeRendererOp
 
+from holoscan.conditions import CountCondition
+from holoscan.operators import FormatConverterOp, VideoStreamRecorderOp
+from holoscan.resources import CudaStreamPool, UnboundedAllocator, BlockMemoryPool, MemoryStorageType
+
+
+FRAME_COUNT = 10
+
 
 class BciVisualizationApp(Application):
     """Kernel Flow BCI Real-Time Reconstruction and Visualization
@@ -65,6 +72,9 @@
         super().__init__(argv, *args, **kwargs)
 
     def compose(self):
+
+        recorder_config = self.kwargs("recorder")
+
         # Resources
         volume_allocator = UnboundedAllocator(self, name="allocator")
         cuda_stream_pool = CudaStreamPool(
@@ -89,6 +99,7 @@
         stream_operator = StreamOperator(self,
             name="stream_operator",
             stream=self._stream,
+            condition=CountCondition(self, count=FRAME_COUNT)
         )
 
         build_rhs_operator = BuildRHSOperator(
@@ -136,6 +147,8 @@
             cuda_stream_pool=cuda_stream_pool,
             **volume_renderer_kwargs,
         )
+        # Eventually ensure deadlock and app exist after many no-op ticks at init
+        volume_renderer.add_arg(CountCondition(self, count=FRAME_COUNT*10))
 
         # IMPORTANT changes to avoid deadlocks of volume_renderer and holoviz
         # when running in multi-threading mode
@@ -153,6 +166,7 @@
             window_title="Kernel Flow BCI Real-Time Reconstruction and Visualization",
             enable_camera_pose_output=True,
             cuda_stream_pool=cuda_stream_pool,
+            enable_render_buffer_output=recorder_config.get("enabled", False),
         )
 
         # ========== Connect Reconstruction Pipeline ==========
@@ -224,6 +238,32 @@
         self.add_flow(color_buffer_passthrough, holoviz, {("color_buffer_out", "receivers")})
         self.add_flow(holoviz, volume_renderer, {("camera_pose_output", "camera_pose")})
 
+        # ========== Recording Pipeline (for testing) ==========
+        if recorder_config.get("enabled",True):
+
+            # Recording pipeline: holoviz -> format_converter -> recorder
+            recorder_pool = BlockMemoryPool(
+                self,
+                name="recorder_pool",
+                storage_type=MemoryStorageType.DEVICE,
+                block_size=1024 * 768 * 4 * 4,  # RGBA float32
+                num_blocks=2,
+            )
+            recorder_format_converter = FormatConverterOp(
+                self,
+                name="recorder_format_converter",
+                in_dtype="rgba8888",
+                out_dtype="rgb888",
+                pool=recorder_pool,
+            )
+            recorder = VideoStreamRecorderOp(self,
+                name="recorder",
+                directory=recorder_config.get("directory", "/tmp"),
+                basename=recorder_config.get("basename", "SOURCE_VIDEO_BASENAME"),
+            )
+            self.add_flow(holoviz, recorder_format_converter, {("render_buffer_output", "source_video")})
+            self.add_flow(recorder_format_converter, recorder)
+
 
 def main():
     parser = argparse.ArgumentParser(
